File: clients/dotnet/tests/Novamem.Tests/ScenarioTests.cs
// The shared behaviour suite (clients/contract/scenarios.json, ADR 0009).
#nullable enable

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace Novamem.Tests;

public sealed class ScenarioServer : IDisposable
{
    public readonly string Url;
    public readonly string Closed;
    readonly Process _p;

    public static string Contract(
        [System.Runtime.CompilerServices.CallerFilePath] string here = ""
    ) => Path.GetFullPath(Path.Combine(Path.GetDirectoryName(here)!, "../../../contract"));

    public ScenarioServer()
    {
        _p = Process.Start(
            new ProcessStartInfo("sh", "scenario-server.sh -scenarios scenarios.json")
            {
                WorkingDirectory = Contract(),
                RedirectStandardOutput = true,
            }
        )!;
        var parts = _p.StandardOutput.ReadLine()!.Split(' ');
        Url = parts[1];
        Closed = parts[2].Split('=')[1];
    }

    public void Dispose() => _p.Kill(true);
}

public sealed class ScenarioTests(ScenarioServer srv) : IClassFixture<ScenarioServer>
{
    static readonly JsonElement Scen = JsonDocument
        .Parse(File.ReadAllText(Path.Combine(ScenarioServer.Contract(), "scenarios.json")))
        .RootElement;
    static string Token => Scen.GetProperty("token").GetString()!;

    public static IEnumerable<object[]> Ids() =>
        Scen.GetProperty("scenarios")
            .EnumerateArray()
            .Select(s => new object[] { s.GetProperty("id").GetString()! });

    static bool IsEmpty(JsonElement r) =>
        (r.ValueKind == JsonValueKind.Array && r.GetArrayLength() == 0)
        || (
            r.ValueKind == JsonValueKind.Object
            && r.TryGetProperty("results", out var res)
            && res.ValueKind == JsonValueKind.Array
            && res.GetArrayLength() == 0
        );

    static string Subset(JsonElement want, JsonElement? got, string path = "$")
    {
        if (want.ValueKind == JsonValueKind.Object)
        {
            if (got is not { ValueKind: JsonValueKind.Object } g)
                return $"{path}: want an object, got {got}";
            foreach (var p in want.EnumerateObject())
            {
                var d = Subset(
                    p.Value,
                    g.TryGetProperty(p.Name, out var v) ? v : null,
                    $"{path}.{p.Name}"
                );
                if (d.Length > 0)
                    return d;
            }
            return "";
        }
        // Raw-text comparison: JsonElement.DeepEquals is .NET 9+, and these
        // expectations are small scalars written the way the SDK writes them.
        return got is { } x && x.GetRawText() == want.GetRawText()
            ? ""
            : $"{path}: got {got}, want {want}";
    }

    // proved by: removing the degraded-empty check in Client.SearchAsync fails
    // search-degraded-empty-is-unavailable; removing the redaction fails
    // token-echoed-in-401-is-redacted.
    [Theory]
    [MemberData(nameof(Ids))]
    public async Task TestScenarios(string id)
    {
        var s = Scen.GetProperty("scenarios")
            .EnumerateArray()
            .First(x => x.GetProperty("id").GetString() == id);
        var call = s.GetProperty("call");
        JsonElement? result = null;
        Exception? err = null;
        if (call.GetProperty("class").GetString() == "ctor")
        {
            var a = call.GetProperty("args");
            try
            {
                _ = new Client(
                    new NovamemOptions
                    {
                        BaseUrl = a.GetProperty("baseUrl")
                            .GetString()!
                            .Replace("<server>", srv.Url),
                        Token = a.GetProperty("token").GetString()!,
                    }
                );
            }
            catch (Exception e)
            {
                err = e;
            }
        }
        else
        {
            var baseUrl =
                s.TryGetProperty("respond", out var rsp) && rsp.GetRawText().Contains("\"refused\"")
                    ? $"http://127.0.0.1:{srv.Closed}/s/{id}"
                    : $"{srv.Url}/s/{id}";
            var o = new NovamemOptions
            {
                BaseUrl = baseUrl,
                Token = Token,
                Timeout = TimeSpan.FromMilliseconds(Scen.GetProperty("timeoutMs").GetInt32()),
            };
            var clients = new Clients(new Client(o), new Management(o), new Admin(o));
            using var cts = new CancellationTokenSource();
            if (call.TryGetProperty("cancelAfterMs", out var c))
                cts.CancelAfter(c.GetInt32());
            try
            {
                var r = await Dispatch.Table[call.GetProperty("method").GetString()!](
                    clients,
                    call.GetProperty("args"),
                    cts.Token
                );
                result = JsonSerializer.SerializeToElement(
                    r,
                    r?.GetType() ?? typeof(object),
                    Json.Options
                );
            }
            catch (Exception e)
            {
                err = e;
            }
        }
        var outcome = err switch
        {
            null => result is { } j && IsEmpty(j) ? "empty" : "ok",
            OperationCanceledException => "canceled",
            NovamemException { IsUnavailable: true } => "unavailable",
            NovamemException { IsNotFound: true } => "not_found",
            _ => "error",
        };
        var exp = s.GetProperty("expect");
        Assert.True(
            exp.GetProperty("outcome").GetString() == outcome,
            $"{id}: outcome {outcome} ({err}), want {exp.GetProperty("outcome")}"
        );
        if (err is not null)
        {
            Assert.DoesNotContain(Token, err.ToString());
            if (err is NovamemException ne)
            {
                if (exp.TryGetProperty("retryable", out var r))
                    Assert.Equal(r.GetBoolean(), ne.IsRetryable);
                if (exp.TryGetProperty("statusCode", out var sc))
                    Assert.Equal(sc.GetInt32(), ne.StatusCode);
                if (exp.TryGetProperty("code", out var code))
                    Assert.Equal(code.GetString(), ne.Code);
            }
            if (exp.TryGetProperty("messageContains", out var mc))
                Assert.Contains(mc.GetString()!, err.Message, StringComparison.OrdinalIgnoreCase);
        }
        if (exp.TryGetProperty("result", out var want))
            Assert.Equal("", Subset(want, result));
        if (call.GetProperty("class").GetString() == "ctor")
            return;
        using var http = new HttpClient();
        var v = await http.GetFromJsonAsync<JsonElement>($"{srv.Url}/_verdict/{id}");
        Assert.True(
            v.GetProperty("mismatches").GetArrayLength() == 0,
            $"{id}: mismatches {v.GetProperty("mismatches")}"
        );
        if (
            s.TryGetProperty("expectRequest", out var er)
            && er.ValueKind == JsonValueKind.Array
            && er.GetArrayLength() == 0
        )
            Assert.Equal(0, v.GetProperty("requests").GetInt32());
    }
}

File: clients/swift/Tests/NovamemTests/ScenarioTests.swift
// The shared behaviour suite (clients/contract/scenarios.json, ADR 0009).

import Foundation
@testable import Novamem
import XCTest
#if canImport(FoundationNetworking)
    import FoundationNetworking
#endif

final class ScenarioTests: XCTestCase {
    static let contract = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        .appendingPathComponent("../../../contract").standardizedFileURL

    func startServer() throws -> (Process, String, String) {
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/bin/sh")
        p.arguments = ["scenario-server.sh", "-scenarios", "scenarios.json"]
        p.currentDirectoryURL = Self.contract
        let pipe = Pipe()
        p.standardOutput = pipe
        try p.run()
        var line = Data()
        while true {
            let b = pipe.fileHandleForReading.readData(ofLength: 1)
            if b.isEmpty || b == Data([0x0A]) {
                break
            }
            line.append(b)
        }
        let parts = String(decoding: line, as: UTF8.self).split(separator: " ").map(String.init)
        return (p, parts[1], String(parts[2].split(separator: "=")[1]))
    }

    func isEmpty(_ d: Data) -> Bool {
        let o = try? JSONSerialization.jsonObject(with: d, options: [.fragmentsAllowed])
        if let a = o as? [Any] {
            return a.isEmpty
        }
        if let m = o as? [String: Any], let r = m["results"] as? [Any] {
            return r.isEmpty
        }
        return false
    }

    func subset(_ want: Any, _ got: Any?, _ path: String = "$") -> String {
        if let w = want as? [String: Any] {
            guard let g = got as? [String: Any] else { return "\(path): want an object, got \(String(describing: got))" }
            for (k, v) in w {
                let d = subset(v, g[k], "\(path).\(k)")
                if !d.isEmpty {
                    return d
                }
            }
            return ""
        }
        return (want as? NSObject)?.isEqual(got) == true ? "" : "\(path): got \(String(describing: got)), want \(want)"
    }

    /// proved by: removing the degraded-empty check in Client.search fails
    /// search-degraded-empty-is-unavailable; removing the redaction fails
    /// token-echoed-in-401-is-redacted.
    func testScenarios() async throws {
        let scen = try XCTUnwrap(try JSONSerialization
            .jsonObject(with: Data(contentsOf: Self.contract.appendingPathComponent("scenarios.json"))) as? [String: Any])
        let token = try XCTUnwrap(scen["token"] as? String)
        let timeout = try Double(XCTUnwrap(scen["timeoutMs"] as? Int)) / 1000
        let (proc, url, closed) = try startServer()
        defer { proc.terminate() }
        var failures: [String] = []

        for s in try XCTUnwrap(scen["scenarios"] as? [[String: Any]]) {
            let id = try XCTUnwrap(s["id"] as? String)
            let call = try XCTUnwrap(s["call"] as? [String: Any])
            var result: Data?
            var err: Error?
            if call["class"] as! String == "ctor" {
                let a = try XCTUnwrap(call["args"] as? [String: String])
                do { _ = try Client(Config(
                    baseURL: XCTUnwrap(a["baseUrl"]?.replacingOccurrences(of: "<server>", with: url)),
                    token: XCTUnwrap(a["token"])
                )) } catch { err = error }
            } else {
                let respond = try String(
                    decoding: JSONSerialization.data(withJSONObject: s["respond"] ?? [:], options: [.fragmentsAllowed]),
                    as: UTF8.self
                )
                let base = respond.contains("\"refused\"") ? "http://127.0.0.1:\(closed)/s/\(id)" : "\(url)/s/\(id)"
                let cfg = try Config(baseURL: base, token: token, timeout: timeout)
                let c = Clients(client: Client(cfg), management: Management(cfg), admin: Admin(cfg))
                let args = try JSONSerialization.data(withJSONObject: call["args"] ?? [:])
                guard let fn = try dispatchTable[XCTUnwrap(call["method"] as? String)]
                else { failures.append("\(id): no dispatch entry"); continue }
                let task = Task { try await fn(c, args) }
                if let ms = call["cancelAfterMs"] as? Int {
                    Task { try await Task.sleep(nanoseconds: UInt64(ms) * 1_000_000); task.cancel() }
                }
                do { result = try await task.value } catch { err = error }
            }
            let outcome: String = switch err {
            case nil: isEmpty(result ?? Data()) ? "empty" : "ok"
            case is CancellationError: "canceled"
            case let e as NovamemError where e.isUnavailable: "unavailable"
            case let e as NovamemError where e.isNotFound: "not_found"
            default: "error"
            }
            let exp = try XCTUnwrap(s["expect"] as? [String: Any])
            if outcome !=
                exp["outcome"] as! String
            {
                failures.append("\(id): outcome \(outcome) (\(String(describing: err))), want \(exp["outcome"]!)")
            }
            if let e = err {
                if "\(e)".contains(token) || String(reflecting: e).contains(token) {
                    failures.append("\(id): token leaked")
                }
                if let ne = e as? NovamemError {
                    if let r = exp["retryable"] as? Bool, r != ne.isRetryable {
                        failures.append("\(id): retryable \(ne.isRetryable)")
                    }
                    if let sc = exp["statusCode"] as? Int, sc != ne.statusCode {
                        failures.append("\(id): status \(ne.statusCode)")
                    }
                    if let code = exp["code"] as? String, code != ne.code {
                        failures.append("\(id): code \(ne.code)")
                    }
                }
                if let m = exp["messageContains"] as? String,
                   !"\(e)".lowercased().contains(m.lowercased())
                {
                    failures.append("\(id): message \(e)")
                }
            }
            if let want = exp["result"] {
                let got = result.flatMap { try? JSONSerialization.jsonObject(with: $0, options: [.fragmentsAllowed]) }
                let d = subset(want, got)
                if !d.isEmpty {
                    failures.append("\(id): result \(d)")
                }
            }
            if try XCTUnwrap(call["class"] as? String) != "ctor" {
                let (vd, _) = try await URLSession.shared.data(from: XCTUnwrap(URL(string: "\(url)/_verdict/\(id)")))
                let v = try XCTUnwrap(try JSONSerialization.jsonObject(with: vd) as? [String: Any])
                if try !XCTUnwrap((v["mismatches"] as? [Any])?.isEmpty) {
                    failures.append("\(id): mismatches \(v["mismatches"]!)")
                }
                if let er = s["expectRequest"] as? [Any], er.isEmpty,
                   try XCTUnwrap(v["requests"] as? Int) != 0
                {
                    failures.append("\(id): expected no request")
                }
            }
        }
        XCTAssert(failures.isEmpty, "\(failures.count) failure(s):\n" + failures.joined(separator: "\n"))
    }
}
